patroni/ha.py

```diff
@@ -1591,18 +1608,27 @@ def before_shutdown() -> None:
 
         self.state_handler.stop(str(mode_control['stop']), checkpoint=bool(mode_control['checkpoint']),
                                 on_safepoint=self.watchdog.disable if self.watchdog.is_running else None,
-                                on_shutdown=on_shutdown if mode_control['release'] else None,
+                                on_shutdown=(on_shutdown if mode_control['release'] and not demote_cluster_with_archive
+                                             else None),
                                 before_shutdown=before_shutdown if mode == 'graceful' else None,
-                                stop_timeout=self.primary_stop_timeout())
+                                stop_timeout=None if demote_cluster_with_archive else self.primary_stop_timeout())
         self.state_handler.set_role(PostgresqlRole.DEMOTED)
         self.set_is_leader(False)
 
         if mode_control['release']:
             if not status['released']:
-                checkpoint_location = self.state_handler.latest_checkpoint_location() if mode == 'graceful' else None
+                checkpoint_lsn, prev_lsn = self.state_handler.latest_checkpoint_locations() \
+                    if mode == 'graceful' else (None, None)
                 with self._async_executor:
-                    self.release_leader_key_voluntarily(checkpoint_location)
-            time.sleep(2)  # Give a time to somebody to take the leader lock
+                    if mode == 'demote-cluster':
+                        # for cluster demotion we need shutdown checkpoint lsn to be written to optime, not the prev one
+                        self.dcs.update_leader(self.cluster, checkpoint_lsn, None, self._failsafe_config())
+                    else:
+                        self.release_leader_key_voluntarily(prev_lsn)
+            if demote_cluster_with_archive:
+                self._rewind.archive_shutdown_checkpoint_wal(cast(str, archive_cmd))
+            else:
+                time.sleep(2)  # Give a time to somebody to take the leader lock
```

Comment:
Something feels wrong here.
Do we really need to sleep if we demote to standby cluster, when there is no archive_command and leader key is not released?